text/0000-bind-helper.md

```diff
@@ -0,0 +1,114 @@
+- Start Date: 2019-03-20
+- Relevant Team(s): Ember.js, Learning
+- RFC PR: https://github.com/emberjs/rfcs/pull/470
+- Tracking: (leave this empty)
+
+# `{{with-args}}` Helper
+
+## Summary
+
+This RFC introduces a new helper, `{{with-args}}`, to allow clear argument passing for functions in templates.
+
+## Motivation
+
+The current `action` helper has a lot of implicit and confusing behavior that is different than the Octane and post Octane programming model.
+
+To understand the complexity of `action` there are many complex behaviors including:
+
+1. Argument partial application (currying)
+2. `this` context binding
+3. `send` checks for Component and Controllers
+
+At build time the `action` helper currently is passed through an AST transform to explicitly bind the `this` to be deterministic at runtime. This is a private API where the outputted Glimmer is not a 1-1 to the template. Also, the `action` helper is confused and has overlap with the `action` modifier which has similar but slightly different behavior.
+
+Instead of this confusing and overloaded behavior, a new `with-args` helper would be introduced to do partial application (with no need for build time private APIs), and context binding will be done instead using the `@action` decorator in classes.
+
+## Detailed design
+
+The `with-args` helper will take in a function and then the set of arguments that will be partially applied to the function.
+
+Here are some examples of the `with-args` helper and the equivalent JS:
```

Comment:
The `call` examples here are not equivalent, as they bind `this` to the method. `{{with-args` specifically does not do this.

Reply:
We're assuming that the wrapped function will have the correct context wherever it is called, e.g. if you were to pass that outer function to `addEventListener`, `this` would be the correct context (the element of the event handler) and binding it would result in the correct `this` being passed to `call`. Not binding it means it would default to whatever the context of `this` was when the function was defined, I _believe_.